---
title: "Simulation of scRNA-seq data and differential stability genes"
author: "Kevin Wang and Hani Kim"
date: "`r paste0('Initiated on 2020 Oct 31, compiled on ', format(Sys.time(), '%Y %b %d'))`"
output:
  html_document:
    code_folding: hide
    fig_height: 12
    fig_width: 12
    toc: yes
    number_sections: true
    toc_depth: 3
    toc_float: yes
    theme: paper
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary
This script performs simulation of scRNA-seq data and differentially stable genes. We use [Splatter](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1305-0) to generate simulated scRNA-seq data.

# Loading 

## Packages
```{r, message = FALSE}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(dplyr)
  library(splatter)
})
```

## Load 3-class cellbench data
```{r}
load("../../../Data/cellbench_sce_3class.RData", verbose=T)
rm(sce_CELseq2_log)
rm(sce_Dropseq_log)
```

## Source functions
```{r}
source("../../../R/simulateDS.R")
source("../../../R/Cepo.R")
```

# Simulate scRNA-seq data
Note that this part has been run on server (set.seed gives different output)
```{r}
## generate simulation parameters 
cellbch = sce_10x_log
cellbch_sub = cellbch %>% filter_col(celltype %in% c("H1975", "H2228"))
params <- splatEstimate(cellbch_sub)

## simulate scRNA-seq data
set.seed(123)
sim <- splatSimulate(params, 
                     nGenes = 10000, 
                     batchCells = c(100),
                     method = "single") %>% 
  scater::logNormCounts()

## simulate 500 DS genes
sim_logcounts = sim %>% logcounts %>% rm_cols_rows_zeroes()

sim_rowmeans = rowMeans(sim_logcounts)
sim_rowsd = matrixStats::rowSds(sim_logcounts)
sim_nzrowmeans = apply(sim_logcounts, 1, nz_mean)
sim_nzrowsd = apply(sim_logcounts, 1, nz_sd)
sim_nzrowsd[is.na(sim_nzrowsd)] <- 0
sim_nz = apply(sim_logcounts, 1, function(x) sum(x!=0)/length(x))

set.seed(164)
gene_sampling_prob1 = range_transform(sim_nzrowmeans) %>% thres_prob(thres = 0.2)
gene_sampling_prob2 = range_transform(sim_nz) %>% thres_prob(thres = 0.4)
gene_sampling_prob = range_transform(gene_sampling_prob1*gene_sampling_prob2) %>% thres_prob(thres = 0.2)

ds_index = sample(rownames(sim_logcounts), 
                  size = 500,
                  prob = gene_sampling_prob)

### simulation differential stability parameters
param_grid = expand.grid(
  sd_scale=seq(1,3.4,by=0.4),
  add_zero=seq(2,14,by=2)) %>%
  as_tibble() 
param_grid

set.seed(123)
sim_ds = sim_logcounts
labels = c(rep("Group1", ncol(sim_ds)/2), rep("Group2", ncol(sim_ds)/2))
labels = as.factor(labels)

### introduce differential stability genes to simulated data
sim_list <- lapply(1:nrow(param_grid), function(param) {
  for(this_gene in ds_index){
    print(this_gene)
    sim_ds[this_gene, ] = replace_ds_genes_v2(sim_ds[this_gene, ], labels,
                                              mean_scale = 1, 
                                              sd_scale = param_grid[param,][[1]],
                                              add_zero = param_grid[param,][[2]])
  }
  return(sim_ds)
})

save(sim_list, file="../../../Data/201210_simulation_stability_genes.RData")
save(sim_ds, labels, ds_index, param_grid, file="../../../201210_simulation_params.RData")
```

# Session Info
```{r}
sessioninfo::session_info()
```
